pkg/estimator/client/general.go

```diff
@@ -102,6 +93,130 @@ func (ge *GeneralEstimator) maxAvailableReplicas(cluster *clusterv1alpha1.Cluste
 	return int32(maximumReplicas) // #nosec G115: integer overflow conversion int64 -> int32
 }
 
+// MaxAvailableComponentSets (generic estimator) – resourceSummary only.
+func (ge *GeneralEstimator) MaxAvailableComponentSets(_ context.Context, req *ComponentSetEstimationRequest) ([]ComponentSetEstimationResponse, error) {
+	responses := make([]ComponentSetEstimationResponse, len(req.Clusters))
+	for i, cluster := range req.Clusters {
+		maxComponentSets := ge.maxAvailableComponentSets(cluster, req.Components)
+		responses[i] = ComponentSetEstimationResponse{Name: cluster.Name, Sets: maxComponentSets}
+	}
+	return responses, nil
+}
+
+func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.Cluster, components []*workv1alpha2.Component) int32 {
+	// Compute per-set aggregate requirement (e.g. CPU, Memory)
+	setCPU, setMem, setPodNumber := perSetRequirement(components)
+	resourceSummary := cluster.Status.ResourceSummary.DeepCopy()
+	if resourceSummary == nil {
+		return 0
+	}
+
+	// Check pod constraint
+	allowedPods := getAllowedPodNumber(resourceSummary)
+	if allowedPods <= 0 {
+		return 0
+	}
+
+	// Upper bound by pods, if set pod requirement is 0 default to allowedPodNumber
+	if setPodNumber <= 0 {
+		return int32(allowedPods) // #nosec G115: integer overflow conversion int64 -> int32
```

Comment:
This was an edge case I was unsure of how to handle. I can't think of cases where a CRD with components would have zero replicas (or pods) to schedule... unless the user explicitly wants the components to be scaled down to 0, and then scale up on a CRD update, which would defeat the purpose of the initial estimation.